File: competitive-programming/src/edu/algos/li13_priorityqueue_heap/medium/L1094_CarPooling.java
package edu.algos.li13_priorityqueue_heap.medium;

import java.util.Arrays;
import java.util.PriorityQueue;

public class L1094_CarPooling {

	public static void main(String[] args) {
		L1094_CarPooling obj = new L1094_CarPooling();
		
		int[][] trips = new int[][]{{2,1,5},{3,3,7}};
		System.out.println(obj.carPooling(trips, 4));
	}

	/*
	 * Problem: https://leetcode.com/problems/car-pooling/
	 * Companies: Amazon
     * Idea: We have process the trips in start order. Hence sort them on start.
     * All the trips that we are picking, we need to find first ending trip to remove it from car. Hence, load picked trips to pq based on ending.
     * Sorting on end will not work as we don't know which trips will pick up.
     * 
	 * */
	public boolean carPooling(int[][] trips, int capacity) {
        Arrays.sort(trips, (t1_start, t2_start) -> t1_start[1]-t2_start[1]);
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((t1_end, t2_end) -> t1_end[2]-t2_end[2]);
        int currPass = 0;
        for(int[] trip: trips){
            while(!pq.isEmpty() && pq.peek()[2] <= trip[1]) {
                currPass -= pq.remove()[0];
            }
            
            currPass += trip[0];
            pq.add(trip);
            
            if(currPass > capacity) return false;
        }
        
        return true;
    }
}

File: dsa/patterns/17_intervals/python/l1094_car_pooling.py
# Problem: https://leetcode.com/problems/car-pooling/
# Companies: Amazon
# Idea: We have to process the trips in start order. Hence sort them on start.
# All the trips that we are picking, we need to find the first ending trip to remove it from car.
# Hence, load picked trips to a min-heap (priority queue) based on ending.
# Sorting on end alone will not work as we don't know which trips will be picked up.
import heapq

def car_pooling(trips, capacity):
    # sort trips by start location
    trips.sort(key=lambda x: x[1])

    # min-heap based on end location
    pq = []  # each entry: (end, passengers)
    curr_pass = 0

    for passengers, start, end in trips:
        # remove all trips that have already ended
        while pq and pq[0][0] <= start:
            ended_end, ended_pass = heapq.heappop(pq)
            curr_pass -= ended_pass

        # add current trip
        curr_pass += passengers
        heapq.heappush(pq, (end, passengers))

        if curr_pass > capacity:
            return False

    return True


# ------------------ Test (Same as Java main) ------------------

trips = [[2, 1, 5], [3, 3, 7]]
print(car_pooling(trips, 4))  # Expected: False
